src/main_context_channel.rs

```diff
@@ -0,0 +1,442 @@
+                    .as_mut()
+                    .expect("ChannelSource called before Receiver was attached");
+                if (&mut *callback.borrow_mut())(item) == Continue(false) {
+                    return glib_ffi::G_SOURCE_REMOVE;
+                }
+            }
+        }
+    }
+
+    glib_ffi::G_SOURCE_CONTINUE
+}
+
+unsafe extern "C" fn finalize<T>(source: *mut glib_ffi::GSource) {
+    let source = &mut *(source as *mut ChannelSource<T>);
+
+    // Drop all memory we own by taking it out of the Options
+    let _ = source.receiver.take();
+    let _ = source.callback.take();
+    let _ = source.source_funcs.take();
+}
+
+impl<T> ChannelSource<T> {
+    fn new(receiver: mpsc::Receiver<T>, priority: Priority) -> ptr::NonNull<ChannelSource<T>> {
+        unsafe {
+            let source_funcs = Box::new(glib_ffi::GSourceFuncs {
+                check: Some(check::<T>),
+                prepare: Some(prepare::<T>),
+                dispatch: Some(dispatch::<T>),
+                finalize: Some(finalize::<T>),
+                closure_callback: None,
+                closure_marshal: None,
+            });
+
+            let source = glib_ffi::g_source_new(
+                mut_override(&*source_funcs),
+                mem::size_of::<ChannelSource<T>>() as u32,
+            ) as *mut ChannelSource<T>;
+            assert!(!source.is_null());
+
+            {
+                let source = &mut *source;
+                ptr::write(&mut source.receiver, Some(receiver));
+                ptr::write(&mut source.callback, None);
+                ptr::write(&mut source.source_funcs, Some(source_funcs));
+                source.thread_id = get_thread_id();
+                source.ready = AtomicBool::new(false);
+            }
+
+            glib_ffi::g_source_set_priority(mut_override(&(*source).source), priority.to_glib());
+
+            ptr::NonNull::new_unchecked(source)
+        }
+    }
+
+    fn mark_ready(&self) {
+        self.ready.store(true, Ordering::SeqCst);
+        unsafe {
+            glib_ffi::g_source_set_ready_time(mut_override(&self.source), 0);
+        }
+    }
+}
+
+/// A `Sender` that can be used to send items to the corresponding main context receiver.
+///
+/// This `Sender` behaves the same as `std::sync::mpsc::Sender`.
+///
+/// See [`MainContext::channel()`] for how to create such a `Sender`.
+///
+/// [`MainContext::channel()`]: struct.MainContext.html#method.channel
+#[derive(Clone, Debug)]
+pub struct Sender<T>(Option<mpsc::Sender<T>>, Source);
+
+impl<T> Sender<T> {
+    /// Sends a value to the channel
+    pub fn send(&self, t: T) -> Result<(), mpsc::SendError<T>> {
+        // If the source is destroyed the receiver end is disconnected
+        if self.1.is_destroyed() {
+            return Err(mpsc::SendError(t));
+        }
+
+        let sender = self.0.as_ref().expect("No Sender anymore");
+        sender.send(t)?;
+
+        // Once sending succeeded, wake up the receiver
```

Comment:
"Once **send** succeeded" maybe?

Reply:
That's not a valid English sentence :)

Reply:
I was thinking about the "send action". English doesn't make sense anyway. :p

Reply:
"the send action" is "sending"

Reply:
Well, I'll trust you on this one. :)